### PR TITLE
Trailing comma after rest - The final fix

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Tracking a comma after a spread/rest element is ticky, because we don't know when we should reset the state. We can't do it in `parseExpression` because, for example, in `[[...a,]] = []` `parseExpression` is finished after parsing the inner array.

### CODE REVIEW

File: packages/babel-parser/src/parser/lval.js [hunk 1]
```diff
@@ -60,6 +60,14 @@ export default class LValParser extends NodeUtils {
             const prop = node.properties[i];
             const isLast = i === last;
             this.toAssignableObjectExpressionProp(prop, isBinding, isLast);
+
+            if (
+              isLast &&
+              prop.type === "RestElement" &&
+              node.extra?.trailingComma
+            ) {
+              this.raiseRestNotLast(node.extra.trailingComma);
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/lval.js]: By raising error at comma position we are not really consistent with where this error is throw in other situation, i.e. in `toAssignableObjectExpressionProp`.

https://github.com/babel/babel/blob/11fa2461cec930168fd43cfeda733aab056d8318/packages/babel-parser/src/parser/lval.js#L132-L134

Hence we could throw at `prop.start`.

File: packages/babel-parser/src/parser/expression.js [hunk 2]
```diff
@@ -1488,7 +1489,11 @@ export default class ExpressionParser extends LValParser {
         first = false;
       } else {
         this.expect(tt.comma);
-        if (this.eat(tt.braceR)) break;
+        if (this.match(tt.braceR)) {
+          this.addExtra(node, "trailingComma", this.state.lastTokStart);
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/expression.js]: What if we instead record the `traillingCommaAttachedNodeStart`? So we can always throw at the node.

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js]: I'd prefer to make `trailingComma` a boolean, and then throw at `props[props.length - 1].start`

**Comment by nicolo-ribaudo** [comment]
@JLHwung I ended up fixing the error location in a different way.
It now throws two different errors:
1. `Rest element must be last element`, pointing to the rest element
1. `Unexpected trailing comma after rest element`, pointing to the trailing comma

File: packages/babel-parser/src/parser/lval.js [hunk 3]
```diff
@@ -455,19 +469,22 @@ export default class LValParser extends NodeUtils {
...
     if (this.match(tt.comma)) {
-      this.raiseRestNotLast(this.state.start);
-    }
-  }
-
-  checkCommaAfterRestFromSpread(): void {
-    if (this.state.commaAfterSpreadAt > -1) {
-      this.raiseRestNotLast(this.state.commaAfterSpreadAt);
+      // TODO: Use lookaheadCharCode after that https://github.com/babel/babel/pull/10371 is merged
+      if (this.lookahead().type == close) {
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/lval.js]: https://github.com/babel/babel/pull/10371

**Pull request merged**